### ISSUE TITLE
Improve information with "Something was left on the Lua stack, this is a bug!" warning

### ISSUE THREAD

**Issue by blueyed** [awesomeWM/awesome#3291]
1. While adding some `print`s to debug timers I've seen this in the log:

```
=== timeout: table: 0x411d6f00 (0.1)
stack traceback:
	/usr/share/awesome/lib/gears/timer.lua:96: in function </usr/share/awesome/lib/gears/timer.lua:95>
2021-03-19 16:56:23 W: Trying to emit signal 'request::select' on non-object
my_activate_filter	window/client(tf /tmp/awesome:2.lo...): 0x402f6db0	ewmh (string)	table: 0x40f4c638
  raise : true (boolean)
2021-03-19 16:56:23 W: awesome: a_glib_poll:454: Something was left on the Lua stack, this is a bug!
-------- Lua stack dump ---------
1: userdata	#48	0x40eeb570
------- Lua stack dump end ------
```

I think it would be useful if there was more information with the warning about the leftovers on the stack, e.g. a (Lua) traceback?

Code ref: https://github.com/awesomeWM/awesome/blob/413d47d5a5b6b9d72b6ea01839601efb092ec583/awesome.c#L452-L457

2. In this case it might be related to the previous warning ('W: Trying to emit signal 'request::select' on non-object'), since there is an older entry there also:
```
2021-03-18 01:37:19 W: Trying to emit signal 'request::select' on non-object
my_activate_filter      window/client([personal:1.1:tmux] ...): 0x40f05e38      ewmh (string)   table: 0>
  raise : true (boolean)
2021-03-18 01:37:19 W: awesome: a_glib_poll:454: Something was left on the Lua stack, this is a bug!
-------- Lua stack dump ---------
1: userdata     #48     0x41b3bd28
------- Lua stack dump end ------
```

Related: #2832

awesome v4.3-900-g4319b1611-dirty (Too long)
 • Compiled against Lua 5.1.4 (running with LuaJIT 2.0.5)
 • API level: 4
 • D-Bus support: yes
 • xcb-errors support: yes
 • execinfo support: yes
 • xcb-randr version: 1.6
 • LGI version: 0.9.2
 • Transparency enabled: yes
 • Custom search paths: no

**Comment by psychon** [comment]
From the symptoms, I guess this is fixed by #3237 (both `"ewmh"` and `"request::select"` appear there).

> I think it would be useful if there was more information with the warning about the leftovers on the stack,

I'm sorry, but this is all the information that is available.

I'm sure you know it, but: Lua's C API offers a stack. To run a Lua function, you have to push the Lua function and its argument on the stack and can then call it. Similarly, results from the function end up on the stack. The stack only has finite size and instead of checking properly that there is space left, we just check that it is empty when we expect it is empty.

This check was added because "back then" we got weird crashes from AwesomeWM that eventually turned out to be due to the Lua stack being full. This check allows to notice "there is some garbage left" much earlier and hopefully also in cases when there will be no crash (yet), thus making these bugs much easier to find.

> e.g. a (Lua) traceback?

There is no Lua code running. The traceback would thus only show `debug.traceback()` and nothing else.

The check works like this:
- At the beginning of a main loop iteration, the Lua stack is empty (that's what we assume).
- Then we handle some events. All event handling code should balance its operation on the Lua stack and remove everything that it pushed onto the stack.
- At the end of the main loop iteration, we check if the stack is empty. If it is not, then something happened that did not use Lua properly, but we have no idea what it was.

We could make that check more often, e.g. after every X11 event was handled, but the information "some `ClientMessage` event was handled and now the Lua stack is no longer empty" is not really helpful either. And this wouldn't work for timers anyway.

**Comment by blueyed** [comment]
Ok, closing then.
Thanks for the explanation, @psychon, as always! :)
